FK_solver/src/FK_solver.cpp

```diff
@@ -0,0 +1,341 @@
+    pcout << "  Initializing the solution vector" << std::endl;
+    solution_owned.reinit(locally_owned_dofs, MPI_COMM_WORLD);
+    delta_owned.reinit(locally_owned_dofs, MPI_COMM_WORLD);
+
+    solution.reinit(locally_owned_dofs, locally_relevant_dofs, MPI_COMM_WORLD);
+    solution_old = solution;
+  }
+}
+
+void
+FK_solver::assemble_system()
+{
+  const unsigned int dofs_per_cell = fe->dofs_per_cell;
+  const unsigned int n_q           = quadrature->size();
+
+  FEValues<dim> fe_values(*fe,
+                          *quadrature,
+                          update_values | update_gradients |
+                            update_quadrature_points | update_JxW_values);
+
+  FullMatrix<double> cell_matrix(dofs_per_cell, dofs_per_cell);
+  Vector<double>     cell_residual(dofs_per_cell);
+
+  std::vector<types::global_dof_index> dof_indices(dofs_per_cell);
+
+  jacobian_matrix = 0.0;
+  residual_vector = 0.0;
+
+  // Value and gradient of the solution on current cell.
+  std::vector<double>         solution_loc(n_q);
+  std::vector<Tensor<1, dim>> solution_gradient_loc(n_q);
+
+  // Value of the solution at previous timestep (un) on current cell.
+  std::vector<double> solution_old_loc(n_q);
+
+  forcing_term.set_time(time);
+
+  for (const auto &cell : dof_handler.active_cell_iterators())
+    {
+      if (!cell->is_locally_owned())
+        continue;
+
+      fe_values.reinit(cell);
+
+      cell_matrix   = 0.0;
+      cell_residual = 0.0;
+
+      fe_values.get_function_values(solution, solution_loc);
+      fe_values.get_function_gradients(solution, solution_gradient_loc);
+      fe_values.get_function_values(solution_old, solution_old_loc);
+
+      for (unsigned int q = 0; q < n_q; ++q) {
+    // Evaluate coefficients on this quadrature node.
+    const double alpha_loc = alpha.value(fe_values.quadrature_point(q));
+
+    // Define the matrix to store D values and populate it.
+    FullMatrix<double> D_matrix(dim, dim);
+    D.matrix_value(fe_values.quadrature_point(q), D_matrix);
+
+    for (unsigned int i = 0; i < dofs_per_cell; ++i) {
+        for (unsigned int j = 0; j < dofs_per_cell; ++j) {
+
+            cell_matrix(i, j) += fe_values.shape_value(i, q) * 
+                                 fe_values.shape_value(j, q) / deltat * 
+                                 fe_values.JxW(q);
+
+            cell_matrix(i, j) -= alpha_loc * 
+                                fe_values.shape_value(j, q) * 
+                                fe_values.shape_value(i, q) * 
+                                fe_values.JxW(q);
+
+            cell_matrix(i, j) += 2.0 * alpha_loc *
+                                 solution_loc[q] * 
+                                 fe_values.shape_value(j, q) * 
+                                 fe_values.shape_value(i, q) * 
+                                 fe_values.JxW(q);
+
+            Tensor<1, dim> grad_phi_i = fe_values.shape_grad(i, q);
+            Tensor<1, dim> grad_phi_j = fe_values.shape_grad(j, q);
```

Comment:
This is actually fairly inefficient: you are allocating and deallocating two tensors for every element, for every quadrature node, for every pair of local DoFs $(i, j)$. The method `FEValues::shape_grad` returns its result by `const` reference precisely to avoid this unnecessary allocation and deallocation. If you want to alias the result for the sake of readability, you can just use
```cpp
const Tensor<1, dim> &grad_phi_i = fe_values.shape_grad(i, q);
const Tensor<1, dim> &grad_phi_j = fe_values.shape_grad(j, q);
```
(emphasis on the variables being const **references** - no copy is made here, you're just giving a new name to the variables returned by `shape_grad`).